dht/range_streamer.cc

```diff
@@ -91,7 +91,13 @@ range_streamer::get_range_fetch_map(const std::unordered_map<dht::token_range, s
         }
 
         if (!found_source) {
-            throw std::runtime_error(format("unable to find sufficient sources for streaming range {} in keyspace {}", range_, keyspace));
+            auto& ks = _db.local().find_keyspace(keyspace);
+            auto rf = ks.get_replication_strategy().get_replication_factor();
+            if (rf == 1) {
```

Comment:
This needs a comment, it is not trivial.

Also, should this be qualified for replace node operations only? For bootstrap/decommission, it means the a node is down, and we will end up with unnecessary data loss. If the node comes back up, we can bootstrap/decomission without data loss.

Reply:
I will add a comment. 

The "abort the operation and bring back the node" applies to replace operation as well. E.g, we can avoid data lose to abort the replace operation and bring back the dead node then do the replace operation. 

Glauber had the argument that if  the rf  is 1 and a node is down, it is fine to lose data. 

The safest way is to abort the operation and let user decide what to do with the keyspace with rf one. I am not sure we want to do this.

Reply:
No asias, I didn't have the argument that if the *source* node is down it is fine to lose data, which seems to be Avi's concern.

What I said is that if we are using RF=1, and you lost a node, then it is fine to accept data loss in this case: no need to error out.

My understanding about what @avikivity is saying is this:
- let's say you are bootstrapping a new node into an RF=1 keyspace.
- Let's say one of your nodes is temporarily down while you bootstrap.

You will not find a source node and bail, right? In that case you'll end up with unnecessary data loss.

Reply:
I understand the bootstrap case. My point is even if with replace operation, we could avoid unnecessary data loss if we abort the replace operation and user can fix the dead node and bring it back.  It is not obvious for the user that she/he is going to lose data if she/he performs the replace operation with rf one keyspace. We do not even give a big warning about this and give a chance for the user to reconsider.

Reply:
So, to summary, I think we have two choices: 

Choice 1) 
Abort if no sufficient source are found. Ask user to decide if she/he wants to proceed the operation. A new config option is needed. 

Choice 2)
Do not abort  if no sufficient source and  rf == 1 and operation == replacing

What do yo think?

Reply:
Asias, there is no way to prevent data loss in a replace operation with RF=1.
The data is already gone , so what is the user supposed to do?
You will error out, and then what?
You will add an option, and then what? If the user decides not to proceed because of data loss, is the cluster dead forever?

replace with RF=1 should keep going - because the data is already lost.
bootstrap or decommission with RF=1 should not, because the data may not be lost yet.
Depends on whether the failure is temporary or permanent.

Reply:
Not really, the dead node might still have the data. For example, the network is broken, we can fix the network and bring back the dead node, to avoid the replace operation.

Reply:
@asias the replace operation is user-initiated.
If the user initiated the operation, this means they have already decided the node is not recoverable and the data is lost from the point of view of the cluster.

Reply:
They might want to reconsider the decision if they know there is a data lose. They might forget there is a keyspace with rf one if there  are a lot of keyspaces or not be aware the replace operation will lose data at all.

Reply:
Asias, you are overthinking this.

Let's say the user reconsiders the decision now that they know there is a keyspace with RF=1.
What do they do? go back in time?

For the common case of replace, the node is dead already. There is nothing the user can possibly do.

You are proposing making this more complex - with flags and such - because there is a possibility (which I never seen) that the node is not really dead.

Reply:
@glommer the user can upload data from a backup if he chooses todo that.

Reply:
then what difference does it make if the data was streamed to the new node or not?

Look: from the point of view of streaming, if you are RF=1 and you lost a node, there is nothing to replace. What is done about that later has nothing to do with replace.